### PR TITLE
MM-54123 Add group slash command

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Adds the ability to invite user groups via `/invite`

Combines same errors for multiple users into a single message depending on user count.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-54123

#### Screenshots
Screenshot 2023-09-08 at 11 10 40 AM

#### Release Note
```release-note
invite slash command now supports custom user groups
```

### CODE REVIEW

**Comment by sbishel** [comment]
Note, this is built on top of 
https://github.com/mattermost/mattermost/pull/22954

File: server/channels/app/slashcommands/command_invite.go [hunk 1]
```diff
@@ -59,28 +70,127 @@ func (i *InviteProvider) doCommand(a *app.App, c request.CTX, args *model.Comman
...
-				continue
+			userError := i.addUserToChannel(a, c, args, targetUser, targetChannel)
+			if userError == NoError {
+				if args.ChannelId != targetChannel.Id {
+					differentChannels = append(differentChannels, targetUser.Username)
+				}
+			} else if userError == UserNotInTeam {
+				nonTeamUsers = append(nonTeamUsers, targetUser.Username)
+			} else if userError == IsConstrained {
+				teamConstrained = append(teamConstrained, targetUser)
```
Comment (BenCookie95) [inline server/channels/app/slashcommands/command_invite.go:90]: Does this account for both team constrained and channel constrained?

File: server/channels/app/slashcommands/command_invite.go [hunk 2]
```diff
@@ -59,28 +70,127 @@ func (i *InviteProvider) doCommand(a *app.App, c request.CTX, args *model.Comman
...
 	return ""
 }
 
+func (i *InviteProvider) getUsersFromMentionName(a *app.App, mentionName string) []*model.User {
+	userProfile, err := a.Srv().Store().User().GetByUsername(mentionName)
+	if err == nil && userProfile.DeleteAt == 0 {
+		return []*model.User{userProfile}
+	}
+
+	group, appErr := a.GetGroupByName(mentionName, model.GroupSearchOpts{})
```
Comment (BenCookie95) [inline server/channels/app/slashcommands/command_invite.go:181]: Do we need to use `FilterAllowReference` here?

**Review by BenCookie95 — CHANGES_REQUESTED** [review]

File: server/channels/app/slashcommands/command_invite.go (hunk 1 above)
Comment (sbishel) [reply server/channels/app/slashcommands/command_invite.go]: Actually only channel constrained...we don't try to add to team, rather show error, cannot be added. I did update the variable to `channelConstrained`, so less confusing.

File: server/channels/app/slashcommands/command_invite.go (hunk 2 above)
Comment (sbishel) [reply server/channels/app/slashcommands/command_invite.go]: Yep...I believe so....added.

**Review by BenCookie95 — APPROVED** [review]
LGTM!

**Review by hmhealey — APPROVED** [review]
Looks good to me.

Since the logic for each of those error types is a bit repetitive, I tried to figure out if there was a more generic way to aggregate those errors (maybe storing them in a `map[UserError]*model.User`?), but nothing really stuck out to me for that would have a real benefit

**Pull request closed without merging**